### ISSUE TITLE
Supporting library variants: Step 1

### ISSUE THREAD

**Issue by diml** [ocaml/dune#615]
The first step is to add internal support for having several possible values for the `requires`, `archives`, ... fields of libraries based on the selected variants. What needs to be done is:

### `Variant.Rules`

Add a module `Variant.Rules`, based on `Findlib.Rules` with the following signature:

```ocaml
module Rules : sig
  type variant = t
  type 'a t

  val get : 'a t -> variants:Set.t -> 'a list

  val map : 'a t -> f:('a -> 'b) -> 'b t

  val make : (variant * 'a) list -> 'a t
  val of_meta_rules : Meta.Simplified.Rules.t -> string t
end with type variant := t
```

### Use `Variant.Rules.t` in `Lib`

Change the type of the following fields of `Lib.Info.t` from `x` to `x Variant.Rules.t`:

- `requires`
- `archives`
- `plugins`
- `foreign_archives`
- `jsoo_runtime`

Basically we need to add it to every field we want to allow the user to write in an `implement` stanza. `pps` might make sense, though unless we support `per_file` in addition of `per_module`, it's not really useful.

The corresponding fields of `Lib.t` needs to be updated as well.

### Allow to select variants at link time

Several functions need to be updated to take a `Variant.Set.t` argument that represents the set of selected variants:
- `Lib.closure`
- `Exe.build_and_link*`
- some of the functions in `Lib.L`

### Update META file generation

Update the META file generation as described [here](https://github.com/ocaml/dune/issues/136#issuecomment-319999885)

**Comment by rgrinberg** [comment]
To correctly implement `val of_meta_rules : Meta.Simplified.Rules.t -> string t`, should we use the extra functionality for interpreting predicates that's defined in `Findlib` (`Vars.get` basically)? I understand that we'll be using a subset of the predicates feature, but it's possible that some users might attempt to write variant outside of jbuilder with their own META files.

**Comment by diml** [comment]
Well, one thing to decide is how much we want to be compatible with findlib predicates. There are two use cases for such compatibility:

1. allow jbuilder users to specify findlib predicates for non-jbuilder libraries, such as in #208 
2. allow jbuilder packages using variants to be used without jbuilder

I don't think it's really hard to support such compatibility, so I'm in favor of trying it, at least initially and revisit this decision if we realize that it doesn't work well in practice.

Regarding the implementation, I suggest to just move Findlib.Rules to Variant.Rules and use that directly. Rules.make can encode the choices in findlib rules.

**Comment by rgrinberg** [comment]
to confirm, should `Rules.make` allow duplicate `variant` in the list? I assume that yes because `get` can simply return all of them.

**Comment by rgrinberg** [comment]
Or perhaps we should change get:

```ocaml
val get : 'a t -> variants:Set.t -> 'a
```

The reason for is that things like Lib.t's new  requires field:

```ocaml
requires: Deps.t Variant.Rules.t
```

are awkward since you don't really want a list of `Deps.t`

**Comment by diml** [comment]
For `Rules.make`, it will be only used for jbuilder defined libraries, so we can expect that they'll be no duplicate.

> are awkward since you don't really want a list of Deps.t

I suppose we need `Deps.concat` to flatten the list. Looking at this again I think there might be something missing in the signature of `make`. Basically the main dependencies should be under no variant, and each variant will augment the list of dependencies.

**Comment by rgrinberg** [comment]
> Basically the main dependencies should be under no variant, and each variant will augment the list of dependencies.

Hmm, but I think he current `Rules.Variant` can already accommodate this. As long as we duplicate the values between the choices (sounds fine with me). What I think is missing however is a default choice when no variants are selected.

**Comment by diml** [comment]
Not duplicating is better in term of performances: in `instantiate` we are going to map the whole `Rules.t` and if things are duplicated we are going to have to resolve the same names several times.

> What I think is missing however is a default choice when no variants are selected.

Sure. Do you agree that when linking that should be an error?

**Comment by rgrinberg** [comment]
> Sure. Do you agree that when linking that should be an error?

Ack'd. All variants that can be specified mutually, need to be specified for linking as far as I understand.

**Comment by rgrinberg** [comment]
> I suppose we need Deps.concat to flatten the list. Looking at this again I think there might be something missing in the signature of make. Basically the main dependencies should be under no variant, and each variant will augment the list of dependencies.

`Deps.concat` is still a bit awkward since we'd have to call `Variant.Rules.get` first. That is, we wouldn't be able to flatten things in `map`.

Do you think it's possible to just simplify `Deps.t` and get rid of `Simple` constructor? To keep it roughly the same for callers, we can add a function like `val simple : (Loc.t * string) list -> Deps.t'. The `Simple` constructor seems like an optimization, but I'm actually not sure how much it's helping. Hitting the `Direct` case of `resolve_complex_deps` seems like it wouldn't be too different.

Anyway, the above is only useful if we plan to keep `get` list valued. How about we do it the naive way (performance wise) and make the variant `'a` valued. This should solve both of these problems, while leaving room for addressing the performance issue.

Regarding performance, there are other ways of addressing the performance issue of resolve, like passing in memoized version of `~f` in `map`. Also, I'm not yet sure how you'd address the performance issue with `resolve` anyway. Even if you'd call `resolve_user_deps` on the `base` and the variant cases separately, the variant would still calculate the closure which could do a bunch of extra work. I might be not fully understanding something here...

Also, if accessing the base value is important for optimization. Perhaps we should simply available when constructing a rule:

```
val map' : 'a t -> f:('a -> base:('b Lazy.t) -> 'b) -> 'b t
```

The above is just 1 way to do it of course, since it's possible that we'd also need the `'a` value of base as well.

**Comment by diml** [comment]
I guess we can start by doing what's simpler and then bench the result. If there is no performance difference, then we can just keep the simpler code.

**Comment by samoht** [comment]
Sorry to interject in the discussion again, but I don't remember if supporting "variants" in different libraries is out of scope or not. The use-case is if someone wants to add a new esoteric backend, let's say for Xen` without having to patch all the existing libraries in the opam universe and get their author to approve that new backend.

**Comment by rgrinberg** [comment]
@samoht the set of backends will have to be defined upfront. I think this limitation is only necessary for findlib compat however.

**Comment by samoht** [comment]
Ok thanks. I guess we will fix that in step n then :-)

**Comment by diml** [comment]
Regarding findlib compatibility, we don't need to support everything. I.e. it seems fine to me if some features are only available when using dune. What that means is that if you want to put a variant in a separate library, you have to accept that non-jbuilder users won't be able to use it, or alternatively you have to implement the necessary support in findlib and wire it in dune.

> How should optional work with variants? Can we declare individual implementations as (optional)?

Possibly, that seems like a reasonable feature one would want. But initially we should do whatever is easier, and leaving it out seems easier.

> Are we going to allow select dependencies on variant implementations?

I think we can, yes.

> I'm a bit unsure as to how to modify Compile.t for variants support. We can either leave it unchanged and pass Compiles.t Variant.Rules.t everywhere. Or, we can change the fields and their accessors to be _ Variant.Rules.t valued. I'm leaning towards the former, b/c it will allow us to have a value to represent the compile information once all the variants have been selected.

That seems sensible to me.

> Are backends and virtual libraries going to be mutually exclusive? I assume so, for simplicity.

Yes, that seems very reasonable

> Can a library depend on a particular variant of another library? E.g. depending on foo will depend on it in a backend independent way, but depending on something like foo.os will force the os variant. This isn't hard to support, but I'm not sure how to make this findlib compatible. Since we'll that depending on a library essentially forces a predicate.

Same remark about findlib compatibility. Note that allowing a library to force a specific variant seems fine. However, assuming that this variant is automatically selected as soon as you depend on this library is more complicated. It basically mean that the set of selected variants grows as you compute the transitive closure. I don't know if this computation is guaranteed to converge.

> How to handle conflicting variant names. E.g. consider two independent libraries foo and bar with x and y variants. If we write an executable that depends on foo and bar, we can either say thatx,yforce(foo.x bar.x),(foo.y bar.y). Or we can be more flexible and allow to select the variant per dependency. I think being more flexible is nice, since it's not necessarily the case thatxandy` are incompatible. It's probably a bit too complicated for v1 but let's not make this impossible in the future.

Possibly but that is going to make things complicated as well. For instance you might end up depending twice on a library but with a different set of variants.

**Comment by rgrinberg** [comment]
One more question:

Is `Mode.Dict.t` obsolete? I think we can replace it with `Variant.Rules.t`

**Comment by diml** [comment]
That seems right, yes

**Comment by rgrinberg** [comment]
Okay, so to keep things simple for a v1. Let's allow only virtual libraries as dependencies. Backends may only be selected via the `variants` option for executables. I don't see a good use a great use case for depending on a particular backend anyway.

Another question:

The status table for libraries is no longer sufficient. The current implementation is:

```
  ; table   : (string, status) Hashtbl.t
```

But the status can now vary depending on the backend. Should we parameterize this status using `Variant.Rules.t` as well?

What about .cmxs files, is it possible to create shared objects for virtual libraries? 	If not, then I suppose we don't allow `(kind ppx_rewriter)` for virtual libs?

**Comment by diml** [comment]
> But the status can now vary depending on the backend. Should we parameterize this status using Variant.Rules.t as well?

Why does the status depends on the selected variants?

> What about .cmxs files, is it possible to create shared objects for virtual libraries? If not, then I suppose we don't allow (kind ppx_rewriter) for virtual libs?

It's likely to be the same issue for cma and cmxa files. We have two choices to implement this feature:

1. we create one archive for the virtual library and then one additional archive per implementation
2. we create only one archive per implementation

(2) has the disadvantage that non virtual modules are duplicated in all archives, however I'm not entirely sure it is possible to implement (1) in the current world.

Note that we don't use cmxs files for ppx rewriters. So the two problems are independent.

**Comment by rgrinberg** [comment]
This has been implemented (although variants are still in beta)
